pkg/networkservice/mechanisms/directmemif/server.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package directmemif provides server chain element that create connection between two memif interfaces
+package directmemif
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/sdk-vppagent/pkg/networkservice/mechanisms/directmemif/proxy"
+	"github.com/networkservicemesh/sdk-vppagent/pkg/networkservice/vppagent"
+)
+
+type directMemifServer struct {
+	executor serialize.Executor
+	proxies  map[string]proxy.Proxy
+}
+
+// NewServer creates new direct memif server
+func NewServer() networkservice.NetworkServiceServer {
+	return &directMemifServer{
+		executor: serialize.NewExecutor(),
+		proxies:  map[string]proxy.Proxy{},
+	}
+}
+
+func (d *directMemifServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	c := vppagent.Config(ctx)
+	l := len(c.GetVppConfig().GetInterfaces())
+	if l < 2 {
+		return next.Server(ctx).Request(ctx, request)
+	}
+	client := c.GetVppConfig().GetInterfaces()[l-2]
+	endpoint := c.GetVppConfig().GetInterfaces()[l-1]
+	c.GetVppConfig().Interfaces = c.GetVppConfig().GetInterfaces()[:l-2]
+	if client.GetMemif() == nil || endpoint.GetMemif() == nil {
+		return next.Server(ctx).Request(ctx, request)
+	}
+	var err error
+	d.executor.SyncExec(func() {
+		_, ok := d.proxies[request.Connection.Id]
+		if ok {
+			err = errors.New(fmt.Sprintf("proxy with id %v already exsist", request.Connection.Id))
```

Comment:
We are going to be getting re-Requests for Connections... I *think* that those will hit this error, and we probably want to just look for the proxies, and if they are already there, move on.  Does that seem right to you?

Reply:
I've added replacing for the case with re-request.